cpp/oneapi/dal/algo/pca/backend/gpu/finalize_train_kernel_cov_dpc.cpp

```diff
@@ -206,29 +119,47 @@ static train_result<Task> train(const context_gpu& ctx,
 
     const auto sums =
         pr::table2ndarray_1d<Float>(q, input.get_partial_sum(), sycl::usm::alloc::device);
+    sycl::event means_event;
     if (desc.get_result_options().test(result_options::means)) {
-        auto [means, means_event] = compute_means(q, rows_count_global, sums, {});
-        means_event.wait_and_throw();
-        result.set_means(homogen_table::wrap(means.flatten(q), 1, column_count));
+        ONEDAL_PROFILER_TASK(compute_means, q);
+        auto means = pr::ndarray<Float, 1>::empty(q, { column_count }, alloc::device);
+        means_event = pr::means(q, rows_count_global, sums, means, {});
+        result.set_means(homogen_table::wrap(means.flatten(q, { means_event }), 1, column_count));
     }
 
     const auto xtx =
         pr::table2ndarray<Float>(q, input.get_partial_crossproduct(), sycl::usm::alloc::device);
-    auto [cov, cov_event] = compute_covariance(q, rows_count_global, xtx, sums, {});
+    auto cov = pr::ndarray<Float, 2>::empty(q, { column_count, column_count }, alloc::device);
+    sycl::event copy_event;
+    {
+        ONEDAL_PROFILER_TASK(copy_xtx, q);
+        copy_event = copy(q, cov, xtx, { means_event });
+    }
+
+    sycl::event cov_event;
+    {
+        ONEDAL_PROFILER_TASK(compute_covariance_matrix, q);
+        cov_event = pr::covariance(q, rows_count_global, sums, cov, bias, { copy_event });
+    }
```

Comment:
Yet another duplication. I would actually recommend to create a header/source pair with this code.